pkg/p2p/server.go

```diff
@@ -0,0 +1,831 @@
+					log.Warn("registering handler failed", zap.Error(err))
+					return errors.Trace(err)
+				}
+				log.Debug("handler registered", zap.String("topic", task.topic))
+			case taskOnDeregisterHandler:
+				if handler, ok := m.handlers[task.topic]; ok {
+					delete(m.handlers, task.topic)
+					go func() {
+						err := handler.GracefulUnregister(ctx, waitUnregisterHandleTimeout)
+						if err != nil {
+							// This can only happen if `ctx` is cancelled or the workerpool
+							// fails to unregister the handle in time, which can be caused
+							// by inappropriate blocking inside the handler.
+							// We use `DPanic` here so that any unexpected blocking can be
+							// caught in tests, but in the same time we can provide better
+							// resilience in production (`DPanic` does not panic in production).
+							//
+							// Note: Even if `GracefulUnregister` does fail, the handle is still
+							// unregistered, only forcefully.
+							log.L().DPanic("failed to gracefully unregister handle",
+								zap.Error(err))
+						}
+						log.Debug("handler deregistered", zap.String("topic", task.topic))
+						m.acks.RemoveTopic(task.topic)
+						if task.done != nil {
+							close(task.done)
+						}
+					}()
+				}
+			case taskOnRegisterPeer:
+				log.Debug("taskOnRegisterPeer",
+					zap.String("sender", task.sender.GetStreamMeta().SenderId),
+					zap.Int64("epoch", task.sender.GetStreamMeta().Epoch))
+				if err := m.registerPeer(ctx, task.sender, task.clientAddr); err != nil {
+					if cerror.ErrPeerMessageStaleConnection.Equal(err) || cerror.ErrPeerMessageDuplicateConnection.Equal(err) {
+						// These two errors should not affect other peers
+						if err1 := task.sender.Send(ctx, errorToRPCResponse(err)); err1 != nil {
+							return errors.Trace(err)
+						}
+						continue // to handling the next task
+					}
+					return errors.Trace(err)
+				}
+			case taskDebugDelay:
+				log.Info("taskDebugDelay started")
+				select {
+				case <-ctx.Done():
+					log.Info("taskDebugDelay canceled")
+					return errors.Trace(ctx.Err())
+				case <-task.doneCh:
+				}
+				log.Info("taskDebugDelay ended")
+			}
+		}
+	}
+}
+
+func (m *MessageServer) tick(ctx context.Context) {
+	var peersToDeregister []*cdcPeer
+	defer func() {
+		for _, peer := range peersToDeregister {
+			// err is nil because the peers are gone already, so sending errors will not succeed.
+			m.deregisterPeer(ctx, peer, nil)
+		}
+	}()
+
+	m.peerLock.RLock()
+	defer m.peerLock.RUnlock()
+
+	for _, peer := range m.peers {
+		var acks []*p2p.Ack
+		m.acks.Range(peer.PeerID, func(topic Topic, seq Seq) bool {
+			acks = append(acks, &p2p.Ack{
+				Topic:   topic,
+				LastSeq: seq,
+			})
+			return true
+		})
+
+		if len(acks) == 0 {
+			continue
+		}
+
+		peer.metricsAckCount.Inc()
+		err := peer.sender.Send(ctx, p2p.SendMessageResponse{
+			Ack:        acks,
+			ExitReason: p2p.ExitReason_OK,
```

Comment:
Does `ExitReason_OK` mean everything is OK and don't need to exit?

Reply:
Yes